### DISCUSSION TITLE
How are the EF Transactions triggered?

### DISCUSSION THREAD

**Discussion by lefig** [meysamhadeli/booking-microservices#309] [Q&A]
Hi,

Thank you for a great project which I have been following for some time. Its the only one that I know of which combines all the best practices in a relatively simple educational form. I have learnt much thanks to you.

And I was wondering how the methods implementing IDbContext get called and when?

Best wishes

Fig

**Comment by meysamhadeli** [comment]
Hi @lefig, 

I am glad that this project was helpful :)

For using db-context in our handler, I use a [mediatr pipeline behavior](https://github.com/jbogard/MediatR) concept for handling transactions and save data in EF Core. You can follow this [EfTxBehavior](https://github.com/meysamhadeli/booking-microservices/blob/main/src/BuildingBlocks/EFCore/EfTxBehavior.cs) for checking how db-context will triggered after compilation our handler. 

> Note: Behavior in mediatr works like Decorator Pattern and here I use a custom EF Core behavior for handling transaction top of SaveChangeAsync in EF Core. And for each handler this part of code run without rewrite manually SaveChangeAsync for each handler and if something is broken during saving the data, every thing will rollback with this custom behavior, because each save operate as a transaction.
